source/MissionAction.cpp

```diff
@@ -356,32 +420,126 @@ bool MissionAction::CanBeDone(const PlayerInfo &player, const shared_ptr<Ship> &
+		bool checkInstalled = locations.installed;
+		bool checkCargo = locations.cargo;
+		bool checkPresent = locations.present;
+		bool checkAbsent = locations.absent;
+		bool checkParked = locations.parked;
+		bool checkUnparked = locations.unparked;
+		
+		if(!locations.empty)
 		{
-			for(const auto &ship : player.Ships())
-				if(!ship->IsDestroyed())
-				{
-					available += ship->Cargo().Get(it.first);
-					available += ship->OutfitCount(it.first);
-				}
+			// The following if statements swap location checks to true if certain keywords have
+			// been omitted, e.g, not specifying installed or cargo means check both.
+			if(!checkFlag && !checkEscorts) {
+				checkEscorts = true;
+				if(!checkAbsent && !checkParked)
+					checkFlag = true;
+			}
+			if(!checkCargo && !checkInstalled)
+			{
+				checkCargo = true;
+				checkInstalled = true;
+			}
+			if(!checkPresent && !checkAbsent)
+			{
+				checkPresent = true;
+				checkAbsent = true;
+			}
+			if(!checkParked && !checkUnparked)
+			{
+				checkParked = true;
+				checkUnparked = true;
+			}
+		}
+		else if(!count)
+		{
+			// Requiring the player to have 0 of this outfit and not specifying where to look
+			// means all ships and all cargo holds must be checked, even if the ship is disabled, 
+			// parked, or out-of-system.
+			checkFlag = true;
+			checkEscorts = true;
+			checkInstalled = true;
+			checkCargo = true;
+			checkPresent = true;
+			checkAbsent = true;
+			checkParked = true;
+			checkUnparked = true;
 		}
 		else
 		{
-			// Required outfits must be present on able ships in the
-			// player's location (or the respective cargo hold).
-			available += flagship ? flagship->OutfitCount(it.first) : 0;
-			available += boardingShip ? flagship->Cargo().Get(it.first)
+			// If no check locations were specified and the required amount is greater than zero, 
+			// check only the flagship.
+			checkFlag = true;
+			checkCargo = true;
+			checkInstalled = true;
+		}
+		
+		// Now that the check booleans have been handled, check the flagship
+		// if necessary.
+		if(checkFlag)
+		{
+			if(checkCargo)
+				available += boardingShip ? flagship->Cargo().Get(it.first)
 					: CountInCargo(it.first, player);
+			if(checkInstalled)
+				available += flagship ? flagship->OutfitCount(it.first) : 0;
+			
+			// If the required amount is 0 and an outfit has already been found,
+			// then return false.
+			if(!count && available > 0)
+				return false;
 		}
 		
-		if(available < it.second)
-			return false;
 		
-		// If the required count is 0, the player must not have any of the outfit.
-		if(checkAll && available)
+		// If the escorts need to be checked, then iterate through all the player's ships
+		if(checkEscorts)
+		{
+			checkFlag = true;
+			for(const auto &ship : player.Ships())
+			{
+				// Ignore destroyed ships.
+				if(ship->IsDestroyed())
+					continue;
+				// The first ship in list is the flagship, so skip over it.
+				if(checkFlag)
```

Comment:
It might be better to do an explicit comparison between `ship` and `flagship`, in case the flagship is ever not the first ship.

Reply:
How would I go about doing that? Everything I tried didn't seem to work.

Reply:
`ship` is a `shared_ptr<Ship>` while `flagship` is a straight-up pointer to a `Ship`; therefore you either need to compare `ship` to `player.FlagshipPtr()` or `ship.get()` to `flagship`. So, most simply:
```cpp
if(checkFlag && ship.get() == flagship)
{
	checkFlag = false;
	continue;
}
```